### PR TITLE
Edit Products > price settings: set currency position based on store settings

### LINKED ISSUE #1945: Top Performers: product images aren't always shown
**Describe the bug**
For some products under the Top Performers section, the image isn't shown in the product row while the product image exists in the Products tab.

**To Reproduce**

Note: this is not always reproducible, some product images are shown while some aren't.

- On web, place an order with the product that has at least an image
- Launch the app
- Go to the Products tab and make sure the product from step 1 is loaded with an image
- Go to "My store" tab and scroll to the Top Performers section --> the product from step 1 should appear with an image (assuming that there are not many other orders for today)

**Screenshots**

Products tab (note that the "Manga & Comics"! product has an image) | Top Performers (the image isn't shown)
-- | --
IMG_7672 | IMG_7674

**Expected behavior**

The product image should show up under the Top Performers section.

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Mobile Environment**
Please include:
- Device: iPhone 6s Plus
- iOS version: iOS 12.1
- WooCommerce iOS version: 3.8.0.0

**WordPress Environment**
WC version 4.0

### LINKED ISSUE #1956: Products: Currency symbol in wrong position in Price section
**Describe the bug**

In my store's currency (GBP) I expect the currency symbol (`£`) to appear to the left of the price.

Instead, it appears on the right side of the `Price` and `Sale price` fields when I edit them.

**To Reproduce**
1. Go to the Products tab.
2. Select a product.
3. Select "Price."
4. Notice that the currency symbol appears to the right of the price/sale price amount, instead of to the left.

**Screenshots**

**Mobile Environment**
Please include:
- Device: iPhone XS
- iOS version: 13.3.1
- WooCommerce iOS version: 3.8.0.0 TestFlight beta

### PR DESCRIPTION
**Pull request by jaclync**
Fixes #1956

Thanks for the suggestions @pmusolino @shiki ! after a series of unsuccessful attempts to extract the price value, validate, and then reformat the price value with the currency symbol, I tried the approach suggested by @pmusolino to update the currency/price stack view to separate the price input handling from the currency symbol, and it seems to work! (with a small gotcha at the end of the "Changes" section below 👇 )

## Changes

- Added `UIStackView` extension for removing all subviews and adding an array of subviews + basic unit tests
- Added `UnitPosition` enum to `UnitInputViewModel` that indicates where the unit is positioned relative to the input text field
- For `UnitInputViewModel`, passed the `UnitPosition` for the following Product fields:
  - Price settings - regular/sale price: set `UnitPosition` based on wp-admin > WooCommerce > Settings > General
  - Inventory settings - stock quantity: set `UnitPosition` to `none`
  - Shipping settings - height, width, length, weight: set `UnitPosition` to `afterInput`
- In `UnitInputTableViewCell`, updated its stack view's spacing and its subviews based on `UnitPosition`:
  - Unfortunately, I had to set `inputTextField.adjustsFontSizeToFitWidth = false` to turn off automatic font size adjustment to fit width, otherwise the font size shrinks by default probably because the text field inside a `UIStackView` doesn't know its available width. Lemme know if you have any ideas to re-enable this!

## Testing

- Go to the Products tab
- Tap on a simple Product
- Tap on the price settings row --> the currency symbol should be positioned relative to the price based on its store settings under wp-admin > WooCommerce > Settings > General

## Example screenshots/screencast

Screencast (including dynamic type):

simulator

Under the currency position option in wp-admin > WooCommerce > Settings > General:

Screen Shot 2020-03-19 at 2 57 15 PM

left with space | left | right with space | right
-- | -- | -- | --
Simulator Screen Shot - iPhone 11 - 2020-03-19 at 14 45 46 | Simulator Screen Shot - iPhone 11 - 2020-03-19 at 14 46 42 | Simulator Screen Shot - iPhone 11 - 2020-03-19 at 14 47 09 | Simulator Screen Shot - iPhone 11 - 2020-03-19 at 14 47 42

sanity tests on other settings screens that have related code changes:

shipping | inventory
-- | --
Simulator Screen Shot - iPhone 11 - 2020-03-19 at 15 07 46 | Simulator Screen Shot - iPhone 11 - 2020-03-19 at 15 07 50

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/Extensions/UIStackView+Subviews.swift [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import UIKit
+
+extension UIStackView {
+    /// Removes all the arranged subviews in the stack view.
+    ///
+    func removeAllArrangedSubviews() {
+        arrangedSubviews.forEach({ removeArrangedSubview($0) })
```
Comment (shiki) [inline WooCommerce/Classes/Extensions/UIStackView+Subviews.swift]: Unimportant: I was just curious so I tried it out. It looks like you can do this magic too: 

```swift
        arrangedSubviews.forEach(removeArrangedSubview)
```

File: WooCommerce/Classes/ViewRelated/ReusableViews/UnitInputTableViewCell.swift [hunk 2]
```diff
@@ -51,10 +79,12 @@ private extension UnitInputTableViewCell {
 
     func configureTitleLabel() {
         titleLabel.applyBodyStyle()
+        titleLabel.setContentCompressionResistancePriority(.required, for: .horizontal)
     }
 
     func configureInputAndUnitStackView() {
         inputAndUnitStackView.spacing = 6
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/ReusableViews/UnitInputTableViewCell.swift]: Should this use `Constants.stackViewSpacing` too?

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Product+PriceSettingsViewModels.swift [hunk 3]
```diff
@@ -11,43 +11,56 @@ extension Product {
...
 
         let currencyFormatter = CurrencyFormatter()
         let currencyCode = CurrencySettings.shared.currencyCode
         let unit = CurrencySettings.shared.symbol(from: currencyCode)
-        var value = currencyFormatter.formatAmount(regularPrice ?? "", with: unit) ?? ""
-        value = value
-            .replacingOccurrences(of: unit, with: "")
-            .replacingOccurrences(of: regexThousandSeparators, with: "$1", options: .regularExpression)
+        let decimalValue = regularPrice.map { currencyFormatter.convertToDecimal(from: $0) ?? 0 }
+        let value = decimalValue.map { "\($0)" } ?? ""
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Product+PriceSettingsViewModels.swift]: @pmusolino Would you mind double-checking this? It doesn't look like it'll cause a regression of what https://github.com/woocommerce/woocommerce-ios/pull/1802 fixed but I'm not sure.

Comment (shiki) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Product+PriceSettingsViewModels.swift]: If this is good, we can remove the `regexThousandSeparators` constant.

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Product+PriceSettingsViewModels.swift [hunk 4]
```diff
@@ -11,43 +11,56 @@ extension Product {
...
                                   value: value,
                                   placeholder: placeholder,
+                                  unitPosition: currencySettings.currencyUnitPosition,
                                   keyboardType: .decimalPad,
                                   inputFormatter: PriceInputFormatter(),
                                   onInputChange: onInputChange)
     }
 }
+
+private extension CurrencySettings {
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Product+PriceSettingsViewModels.swift]: Noice. 👍

File: WooCommerce/Classes/ViewRelated/ReusableViews/UnitInputTableViewCell.swift [hunk 5]
```diff
@@ -41,6 +50,25 @@ final class UnitInputTableViewCell: UITableViewCell {
...
+        }
+
+        switch viewModel.unitPosition {
+        case .beforeInput, .afterInput:
+            inputAndUnitStackView.spacing = Constants.stackViewSpacing
+        case .afterInputWithoutSpace, .beforeInputWithoutSpace:
+            inputAndUnitStackView.spacing = 0
+        default:
+            break
+        }
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/ReusableViews/UnitInputTableViewCell.swift]: Seeing that this is quite long, it's easy to be misinformed in the future that these are unrelated to each other. I believe this deserves its own function to better express the intent:

```suggestion
        rearrangetInputAndUnitStackViewSubviews(using: viewModel.unitPosition)
    }

    private func rearrangetInputAndUnitStackViewSubviews(using position: UnitInputViewModel.UnitPosition) {
        inputAndUnitStackView.removeAllArrangedSubviews()

        switch position {
        case .beforeInput, .beforeInputWithoutSpace:
            inputAndUnitStackView.addArrangedSubviews([unitLabel, inputTextField])
        case .afterInput, .afterInputWithoutSpace:
            inputAndUnitStackView.addArrangedSubviews([inputTextField, unitLabel])
        case .none:
            inputAndUnitStackView.addArrangedSubviews([inputTextField])
        }

        switch position {
        case .beforeInput, .afterInput:
            inputAndUnitStackView.spacing = Constants.stackViewSpacing
        case .afterInputWithoutSpace, .beforeInputWithoutSpace:
            inputAndUnitStackView.spacing = 0
        default:
            break
        }
    }

```

File: WooCommerce/Classes/Extensions/UIStackView+Subviews.swift (hunk 1 above)
Comment (jaclync) [reply WooCommerce/Classes/Extensions/UIStackView+Subviews.swift]: I've seen this syntax but I think SourceKit wasn't working properly in my Xcode so it didn't get autocomplete suggestions when I tried this before. will update!

**Comment by jaclync** [comment]
thanks for the suggestions, updated the PR for another pass @shiki @pmusolino !

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Product+PriceSettingsViewModels.swift (hunk 3 above)
Comment (pmusolino) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Product+PriceSettingsViewModels.swift]: I confirm that there are no regressions, and the tests that I wrote on the previous PR are green 👍 Well done @jaclync! 😉

Comment (shiki) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Price/Product+PriceSettingsViewModels.swift]: Thank you, Paolo! 🙏

**Comment by jkmassel** [comment]
Howdy – we're freezing `3.9` today, so this PR is being bumped to `4.0`. If needs to be released as part of `4.0`, please re-target it at the `release/4.0` branch.

**Comment by jaclync** [comment]
> I've only discovered a small regression on how the price appears when there are no decimals after the comma/dot. It is not a mistake, but the price does not match how it's in the product detail screen.

good point! I decided to revert the changes in the regular and sale price initial values in https://github.com/woocommerce/woocommerce-ios/pull/2009/commits/dc31d575ea0a4b14b866a324275925258a30691a just so that this PR focus on the addition of unit position

**Pull request merged**